python/pyspark/ml/tuning.py

```diff
@@ -847,9 +850,23 @@ def _fit(self, dataset: DataFrame) -> "CrossValidatorModel":
             subModels = [[None for j in range(numModels)] for i in range(nFolds)]
 
         datasets = self._kFold(dataset)
+
+        tmp_dfs_path = _get_temp_dfs_path()
+        spark_session = SparkSession.getActiveSession()
         for i in range(nFolds):
-            validation = datasets[i][1].cache()
-            train = datasets[i][0].cache()
+            validation = datasets[i][1]
+            train = datasets[i][0]
+
+            if tmp_dfs_path:
```

Comment:
let's define a helper function:
```

def _cache(df):
    if ...:
        df.cache()
    else:
        spark = df._session
        df.save
        spark.read...


```

Reply:
then we need to handler the uncache step together in the helper function, i.e., it should be  a context manager. is this you want ?

Reply:
context manager is a good idea